fastar/core.py

```diff
@@ -149,27 +145,35 @@ def _toposort(self, box):
         (self, to_global_coords(b)) for b in static_box_finder(
             (local_child_counts == 0) & (getbox(self.state, box) != KNOWN), 1)]
     while childless_boxes:
-      arr, box = childless_boxes.pop()
-      sorted_boxes.append((arr, box))
-      invals, _, primitive, params, _ = arr.eqn
-      instarts, counts = backward_rules[primitive](box, *invals, **params)
-      for ival, istart, count in zip(invals, instarts, counts):
-        if isinstance(ival, LazyArray) and istart is not None:
-          ibox = istart, count.shape
-          to_iglobal_coords = lambda b: (np.add(istart, b[0]), b[1])
-          ival.child_counts.subtract(
-              ibox, count * (getbox(ival.state, ibox) != KNOWN))
-          ilocal_child_counts = ival.child_counts.get(ibox)
-          childless_boxes.extend(
-              [(ival, to_iglobal_coords(b))
-               for b in static_box_finder(
-                   (ilocal_child_counts == 0) &
-                   (getbox(ival.state, ibox) != KNOWN), 1)])
+      self.process_childless_box(childless_boxes, sorted_boxes)
     return sorted_boxes[::-1]
 
+  def process_childless_box(self, childless_boxes, sorted_boxes):
```

Comment:
@j-towns If you are wondering why I split this out: It allows to capture variables within the `outslice_from_invals` directly. It would be loop variables otherwise, would need a workaround to avoid change between iterations.

Reply:
I'm a little bit confused by this, but it doesn't make too much difference to have it split out so no problem.

Reply:
It's a workaround for one of Python's meanest gotchas:

```python
fs = [(lambda: a) for a in range(3)]
print([f() for f in fs]) # outputs [2, 2, 2], not [0, 1, 2]
```

Reply:
Oh cool, that makes perfect sense. I've been hit by that one before.